### ISSUE TITLE
cocos2d-x iOS loop effect interrupt by the 32th effect

### ISSUE THREAD

**Issue by huujiee** [cocos2d/cocos2d-x#14335]
http://stackoverflow.com/questions/26945149/there-has-a-bug-with-openal-or-cocos2dx

**Comment by pandamicro** [comment]
@liamcindy Please take a look

**Comment by pandamicro** [comment]
Maybe related to https://github.com/cocos2d/cocos2d-x/issues/14349

**Comment by dumganhar** [comment]
The person on the second floor said nearly correct.

> At the end, I finally find the problem and fix it.
> 
> you need modify the init function in CDAudioManager.m file
> 
> add code:
> 
> [soundEngine setSourceGroupNonInterruptible:0 isNonInterruptible:TRUE];
> after line:
> 
> soundEngine = [[CDSoundEngine alloc] init];
> default cocos2d sound engine don't check with looping or playing sfx. Now it will do it. Why me :< find so hard. Because game warning sfx stop strangely, and then find the sfx play problem, and then go into the black hole (sound engine) :< so sad...

But I don't wanna modify the default behavior of `CocosDeshion` which will interrupt the old audio  when there're more 32 audio files being played at the same time.

Compared to the post, I suggest to invoke `setSourceGroupNoInterruptible` after `[self _setUpSourceGroups:sourceDefs total:1];`

That's:

``` objective-c
- (id)init
{    
    if ((self = [super init])) {

        //Create mutexes
        _mutexBufferLoad = [[NSObject alloc] init];

        asynchLoadProgress_ = 0.0f;

        bufferTotal = CD_BUFFERS_START;
        _buffers = (bufferInfo *)malloc( sizeof(_buffers[0]) * bufferTotal);

        // Initialize our OpenAL environment
        if ([self _initOpenAL]) {
            //Set up the default source group - a single group that contains all the sources
            int sourceDefs[1];
            sourceDefs[0] = self.sourceTotal;
            [self _setUpSourceGroups:sourceDefs total:1];
//////////////////////////////////////////////////////////////
            [self setSourceGroupNonInterruptible:0 isNonInterruptible:true]; // Add this line to modify the default behavior.
//////////////////////////////////////////////////////////////
            functioning_ = YES;
            //Synchronize premute gain
            _preMuteGain = self.masterGain;
            mute_ = NO;
            enabled_ = YES;
            //Test whether get gain works for sources
            [self _testGetGain];
        } else {
            //Something went wrong with OpenAL
            functioning_ = NO;
        }
    }

    return self;
}
```

**Comment by piotrros** [comment]
I just tried latest version from github (3.17) and this issue is still here.

I play one long sound (about 30 seconds) and move character around (it makes a sound each time it moves) and after several tries (I guess 32) long sound stops.

But my fix from here:

https://stackoverflow.com/a/33609360/1264375

still works. Can you add it before release of 3.17?

**Comment by drelaptop** [comment]
@piotrros 3.17 have been tested, and almost to be released, so it couldn't to add into 3.17.

Another side about the issue, I think it been ignored for dumganhar did a response. I will try to test it after 3.17 released. please set me as the Assignees @minggo , the changes will be merged into v3 branch if you are right.
